File: app/src/main/java/com/websarva/wings/android/zuboradiary/data/network/WeatherApiRepository.java
package com.websarva.wings.android.zuboradiary.data.network;

import android.util.Log;

import androidx.annotation.IntRange;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.Objects;

import javax.inject.Inject;

import retrofit2.Call;

public class WeatherApiRepository {
    private final WeatherApiService weatherApiService;
    private final String QUERY_DAIRY_PARAMETER = "weather_code";
    private final String QUERY_TIME_ZONE_PARAMETER = "Asia/Tokyo";
    private final int MIN_PAST_DAYS = 1; //過去天気情報取得可能最小日
    private final int MAX_PAST_DAYS = 92; //過去天気情報取得可能最大日

    @Inject
    public WeatherApiRepository(WeatherApiService weatherApiService) {
        Objects.requireNonNull(weatherApiService);

        this.weatherApiService = weatherApiService;
    }

    public boolean canFetchWeatherInfo(LocalDate date) {
        Objects.requireNonNull(date);

        LocalDate currentDate = LocalDate.now();
        Log.d("fetchWeatherInfo", "isAfter:" + date.isAfter(currentDate));
        if (date.isAfter(currentDate)) return false;

        long betweenDays = ChronoUnit.DAYS.between(date, currentDate);
        Log.d("fetchWeatherInfo", "betweenDays:" + betweenDays);

        return betweenDays <= MAX_PAST_DAYS;
    }

    public Call<WeatherApiResponse> fetchTodayWeatherInfo(GeoCoordinates geoCoordinates) {
        Objects.requireNonNull(geoCoordinates);

        return weatherApiService.getWeather(
                String.valueOf(geoCoordinates.getLatitude()),
                String.valueOf(geoCoordinates.getLongitude()),
                QUERY_DAIRY_PARAMETER,
                QUERY_TIME_ZONE_PARAMETER,
                "0" /*today*/,
                "1" /*1日分*/);
    }

    public Call<WeatherApiResponse> fetchPastDayWeatherInfo(
            GeoCoordinates geoCoordinates, @IntRange(from = MIN_PAST_DAYS, to = MAX_PAST_DAYS) int numPastDays) {
        Objects.requireNonNull(geoCoordinates);
        if (numPastDays < MIN_PAST_DAYS || numPastDays > MAX_PAST_DAYS) throw new IllegalArgumentException();

        return weatherApiService.getWeather(
                String.valueOf(geoCoordinates.getLatitude()),
                String.valueOf(geoCoordinates.getLongitude()),
                QUERY_DAIRY_PARAMETER,
                QUERY_TIME_ZONE_PARAMETER,
                String.valueOf(numPastDays),
                "0" /*1日分(過去日から1日分取得する場合"0"を代入)*/);
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/data/network/WeatherApiRepository.kt
package com.websarva.wings.android.zuboradiary.data.network

import android.util.Log
import androidx.annotation.IntRange
import com.websarva.wings.android.zuboradiary.getLogTag
import retrofit2.Response
import java.time.LocalDate
import java.time.temporal.ChronoUnit

internal class WeatherApiRepository (private val weatherApiService: WeatherApiService) {

    // MEMO:@Suppress("unused")が不要と警告が発生したので削除したが、"unused"警告が再発する。
    //      その為、@Suppress("RedundantSuppression")で警告回避。
    @Suppress("unused", "RedundantSuppression") // MEMO:@IntRangeで使用する為、@Suppressで警告回避。
    companion object {
        private const val MIN_PAST_DAYS = 1 //過去天気情報取得可能最小日
        private const val MAX_PAST_DAYS = 92 //過去天気情報取得可能最大日
    }

    private val logTag = getLogTag()

    private val queryDiaryParameter = "weather_code"
    private val queryTimeZoneParameter = "Asia/Tokyo"


    fun canFetchWeatherInfo(date: LocalDate): Boolean {
        val currentDate = LocalDate.now()

        if (date.isAfter(currentDate)) {
            Log.d(logTag, "canFetchWeatherInfo(date = $date) = false")
            return false
        }

        val betweenDays = ChronoUnit.DAYS.between(date, currentDate)
        val result = betweenDays <= MAX_PAST_DAYS
        Log.d(logTag, "canFetchWeatherInfo(date = $date) = $result")
        return result
    }

    suspend fun fetchTodayWeatherInfo(geoCoordinates: GeoCoordinates): Response<WeatherApiData> {
        return weatherApiService.getWeather(
            geoCoordinates.latitude.toString(),
            geoCoordinates.longitude.toString(),
            queryDiaryParameter,
            queryTimeZoneParameter,
            "0",  /*today*/
            "1" /*1日分*/
        )
    }

    suspend fun fetchPastDayWeatherInfo(
        geoCoordinates: GeoCoordinates,
        @IntRange(from = MIN_PAST_DAYS.toLong(), to = MAX_PAST_DAYS.toLong())
        numPastDays: Int
    ): Response<WeatherApiData> {
        require(numPastDays >= MIN_PAST_DAYS)
        require(numPastDays <= MAX_PAST_DAYS)

        return weatherApiService.getWeather(
            geoCoordinates.latitude.toString(),
            geoCoordinates.longitude.toString(),
            queryDiaryParameter,
            queryTimeZoneParameter,
            numPastDays.toString(),
            "0" /*1日分(過去日から1日分取得する場合"0"を代入)*/
        )
    }
}
